File: 0735-asteroid-collision/0735-asteroid-collision.java
class Solution {
    private static boolean track;
    public int[] asteroidCollision(int[] ast) {
        int[] pow = new int[ast.length];
        int curr_pow = 0;
        for (int s: ast) {
            if (s > 0 || (curr_pow > 0 && pow[curr_pow - 1] < 0)) pow[curr_pow++] = s;
            else {
                track = true;
                while (curr_pow > 0 && pow[curr_pow - 1] > 0) {
                    if (!track) break;
                    if (s + pow[curr_pow - 1] <= 0) {
                        track = !(s == -pow[curr_pow-- - 1]);
                    } else track = false;
                }
                if (track) pow[curr_pow++] = s;
            }
        }
        return Arrays.copyOf(pow, curr_pow);
    }
}

File: 0735-asteroid-collision/0735-asteroid-collision.cpp
class Solution {
public:
    vector<int> asteroidCollision(vector<int>& ast) {
        vector<int> pow;
        bool track;
        for (int s: ast) {
            if (s > 0 || (!pow.empty() && pow.back() < 0)) pow.push_back(s);
            else {
                track = true;
                while (!pow.empty() && pow.back() > 0) {
                    if (!track) break;
                    if (s + pow.back() <= 0) {
                        track = !(s == -pow.back());
                        pow.pop_back();
                    } else track = false;
                }
                if (track) pow.push_back(s);
            }
        }
        return pow;
    }
};

auto init = []() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
    return 'c';
}();